[4.0]  Enabling stemming in com_finder for mono-lingual sites

Hackwar opened the pull request:
This PR extends the work of #20391 and while it can be tested the way it is, #20391 should be merged first and later this here.

This PR adds a new option to the configuration of Finder/Smart Search to select the default stemmer to use for content items marked as "All" or monolingual sites. You can select "None", a specific language or the default language for the frontend, that is set in the language manager.

All the code for this feature is in FinderIndexerHelper::stem().

## How to test
1. Install 4.0 with sample data
2. Run indexer in the backend and see that content set to "All" in the DB in #__finder_terms is not stemmed.
3. Go to component configuration and set the default stemmer to your prefered setting.
4. Clear Index and start Index again, noticing that content set to "All" is now stemmed, too.

Hackwar committed:
Implementing language specific tokeniser and stemmer for com_finder

Hackwar committed:
Fixing comments

Hackwar committed:
Fixing codestyle

Hackwar committed:
Merge branch '4.0-dev' into j4finder_language

Hackwar committed:
Merge branch '4.0-dev' into j4finder_language

Hackwar committed:
Merge branch '4.0-dev' into j4finder_language

Hackwar committed:
Merge branch '4.0-dev' into j4finder_language

Hackwar committed:
Use full language identifier for suggestions in Finder

Hackwar committed:
Merge branch '4.0-dev' into j4finder_language

Hackwar committed:
Enabling stemming in com_finder for mono-lingual sites

brianteeman reviewed (approved):
approve strings

infograf768 commented:
As repeated multiple times in https://github.com/joomla/joomla-cms/pull/20391, we do not approve adding the stemmers in the language packs, which makes this PR here unacceptable.

brianteeman commented:
Please change we to I - you do not speak for everyone

infograf768 commented:
I am not alone as you know well but let's have a joke: Never heard of "We the King" ?

brianteeman commented:
I believe you mean the "Royal We" https://en.wikipedia.org/wiki/Royal_we

infograf768 commented:
Specially to this part though as I switched recently to a North Indian status :
> in Hindustani and other North Indian languages, the majestic plural is a common way for elder speakers to refer to themselves when addressing those younger than them.

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4finder_language

Hackwar committed:
Adding a bunch of stemmers to com_finder

Hackwar committed:
Fixing variable names in finder language stemmer files

Hackwar committed:
Codestyle

Hackwar committed:
Merge branch '4.0-dev' into j4finder_language

Hackwar committed:
Removing test files of php-stemmer

Hackwar committed:
Merge branch 'j4finder_language' of github.com:Hackwar/joomla-cms into j4finder_language

Commit:
Fixing the Greek stemmer

Importing non-working stemmer from https://gist.github.com/Hackwar/07d51d81899da96fc0d785c61ed3178f

Commit:
Fixing the Greek stemmer

First pass, using correct character set

Commit:
Fixing the Greek stemmer

Removing variables to indicate character encoding change. This is not
required in Joomla!. The original code was transliterating from UTF-8
to ISO-8859-7 to solve strlen() issues. However, Joomla! requires
mbstring to be installed, therefore we can use the mb_* functions to
correctly count characters.

Commit:
Fixing the Greek stemmer

Use mb_strlen to avoid issued with multibyte string length calculations.

Commit:
Fixing the Greek stemmer

Code style

Commit:
Fixing the Greek stemmer

Fix return type, remove unused debug variable

Commit:
Fixing the Greek stemmer

Updated copyright headers

Commit:
Fixing the Greek stemmer

Remove unused variables

Hackwar committed:
Merge branch '4.0-dev' into j4finder_stemmonolingual

Hackwar committed:
Merge pull request #23 from nikosdion/feature/j4finder_language

Updated Greek stemmer

Hackwar committed:
Merge branch '4.0-dev' into j4finder_language

Commit:
Greek stemmer fix

Address conversion to / from uppercase correctly. The previous code
was hinged on the internal byte representation of the ISO-8859-7
encoding and the way PHP works inconsistently with strings, handling
them as byte-indexed arrays. This is dangerous stuff and SHOULD break
in PHP 8. This updated code uses the explicit mb_string methods and
proper UTF-8 encoding to prevent nasty surprises.

Commit:
Greek stemmer fix

Tokenizer fixes. Greek uses special punctuation such as several forms
of calligraphic / typographic single and double quotes, middle dot
(ano teleia), ellipsis, cross (dagger), star, paragraph and the plus/
minus glyph. These were incorrectly interpreted as letter glyphs by
the generic tokenizer resulting in problematic stemming of words
adjacent to this punctuation.

Commit:
Greek stemmer fix

Rebasing to the LGPLv2.1 Drupal stemmer.

So, here's the funny thing. The only part of the code which I kept is,
in fact, identical to the LGPLv2.1 version of the code. The only parts
which were different between the two versions (LGPLv2.1 and GPLv3) of
the Drupal stemmer are irrelevant since they BOTH require ISO-8859-7
encoding which is incompatible with Joomla! anyway. I have written my
own case converter for the stemmer which is of course licensed under
whatever I want. Since I've signed the Joomla! Contributor Agreement
all those years ago I am licensing it "GPLv2 or later" and
contributing it to Joomla!. Licensing problem solved!

Hackwar committed:
Merge pull request #24 from nikosdion/feature/j4finder_language

Fixing the Greek stemmer

Commit:
Fixing the Greek stemmer

Fixing monkey-copy of rule application code by previous developers.

Implement a missing rule which was partially added by previous devs.

Refactored the static array into the only method that's using it.

Missing since DocBlock tags.

Hackwar committed:
Merge pull request #25 from nikosdion/feature/j4finder_language

Fixing the Greek stemmer

Hackwar committed:
Changing language classes back to primary language selection in com_finder

Hackwar committed:
com_finder: simplifying the storage code in the tokeniser cache

Hackwar committed:
Adding encoding parameter to phpcs

Hackwar committed:
Codestyle

Hackwar committed:
Adding encoding to phpcs in .drone.yml

Hackwar committed:
Merge branch 'j4finder_language' of github.com:Hackwar/joomla-cms into j4finder_stemmonolingual

Hackwar committed:
Applying default language also to tokenisation

Hackwar committed:
Merge branch 'j4finder_stemmonolingual' of github.com:Hackwar/joomla-cms into j4finder_stemmonolingual

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4finder_language

Hackwar committed:
Fixing required terms and suggestions

Hackwar committed:
Merge branch 'j4finder_language' of github.com:Hackwar/joomla-cms into j4finder_stemmonolingual

Hackwar committed:
Codestyle

Hackwar committed:
Merge branch 'j4finder_language' of github.com:Hackwar/joomla-cms into j4finder_stemmonolingual

Hackwar commented:
I've now extended this PR to set the right language for tokenisation. Still waiting for the stemmer PR to be merged. That would clean up this one here a lot...

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4finder_stemmonolingual

# Conflicts:
#	administrator/components/com_finder/config.xml
#	administrator/components/com_finder/helpers/indexer/helper.php
#	administrator/language/en-GB/en-GB.com_finder.ini
#	components/com_finder/Model/SearchModel.php
#	composer.lock
#	libraries/vendor/composer/autoload_classmap.php
#	libraries/vendor/composer/autoload_psr4.php
#	libraries/vendor/composer/autoload_static.php

Hackwar committed:
Codestyle

Quy commented on administrator/language/en-GB/en-GB.com_finder.ini:
```diff
@@ -35,6 +35,10 @@ COM_FINDER_CONFIG_SORT_OPTION_LIST_PRICE="List price"
 COM_FINDER_CONFIG_SORT_OPTION_RELEVANCE="Relevance"
 COM_FINDER_CONFIG_SORT_OPTION_START_DATE="Date"
 COM_FINDER_CONFIG_SORT_ORDER_LABEL="Sort Field"
+COM_FINDER_CONFIG_LANGUAGE_DEFAULT_LABEL="Default Language"
```
Please sort keys in alpha order.

Hackwar commented:
Ok, so merging the stemmer PR in here did not clean up the number of commits... :man_facepalming: But it did clean up the change-delta, so it should be pretty obvious what I'm doing here. This is ready to test and to commit. :wink:

Quy commented on administrator/language/en-GB/en-GB.com_finder.ini:
```diff
@@ -35,6 +35,10 @@ COM_FINDER_CONFIG_SORT_OPTION_LIST_PRICE="List price"
 COM_FINDER_CONFIG_SORT_OPTION_START_DATE="Date"
 COM_FINDER_CONFIG_SORT_ORDER_LABEL="Sort Field"
+COM_FINDER_CONFIG_LANGUAGE_DEFAULT_LABEL="Default Language"
+COM_FINDER_CONFIG_LANGUAGE_DEFAULT_DESC="Set the language to be used for non-multilingual sites or content marked as \"All\". (Depends on available languages.)"
```
String not used.

Hackwar replied:
fixed

Hackwar committed:
Soorting translation strings alphanumerically

carlitorweb commented:
I have tested this item :white_check_mark: successfully on fb32ca0ef5efc8040fa8c4c4f756e62f6f435390<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20562">issues.joomla.org/tracker/joomla-cms/20562</a>.</sub>

infograf768 commented:
`COM_FINDER_CONFIG_LANGUAGE_DEFAULT_DEFAULT_LANGUAGE="Default installed language"`

In fact it is `"Default Site Language"` if I do not mistake.

Also (unhappily), the description does not display (grrr)
String exists: 
`COM_FINDER_CONFIG_LANGUAGE_DEFAULT_DESC="Set the language to be used for non-multilingual sites or content marked as \"All\". (Depends on available languages.)"`
but it is not added in the field and, if I add it, it does not display.
The label is quite insufficient to understand what this parameter is about... 
(If solved, I would anyway take out `(Depends on available languages.)` as it is quite obvious.

In any case, the field type should NOT be `contentlanguage` as these can be deleted on a monolanguage site, but simply `language`.

infograf768 commented:
I have tested this item :red_circle: unsuccessfully on fb32ca0ef5efc8040fa8c4c4f756e62f6f435390<br /><br />Just because of the field type. Otherwise it works fine afaics<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20562">issues.joomla.org/tracker/joomla-cms/20562</a>.</sub>

Hackwar committed:
Changing field type and language strings

Hackwar commented:
Changed and fixed.

infograf768 commented:
I have tested this item :white_check_mark: successfully on fb32ca0ef5efc8040fa8c4c4f756e62f6f435390<br /><br />OK for me. People fighting against Tips in 4.0 should really think better.

Such a parameter does need explanations and not through a possible doc in a unique language. The description of the field should display here.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20562">issues.joomla.org/tracker/joomla-cms/20562</a>.</sub>

mbabker commented:
The fight isn’t against tips, full stop.  The fight is against tips by
default on all the things and tips in an inaccessible manner.  I don’t
think it’s too much to ask to let people figure out a solution to that
instead of retaining the status quo because it’s easy…

On Thu, Jun 21, 2018 at 8:17 AM infograf768 <notifications@github.com>
wrote:

> I have tested this item ✅ successfully on fb32ca0
> <https://github.com/joomla/joomla-cms/commit/fb32ca0ef5efc8040fa8c4c4f756e62f6f435390>
>
> OK for me. People fighting against Tips in 4.0 should really think better.
>
> Such a parameter does need explanations and not through a possible doc in
> a unique language. The description of the field should display here.
> ------------------------------
> This comment was created with the J!Tracker Application
> <https://github.com/joomla/jissues> at
> issues.joomla.org/tracker/joomla-cms/20562.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/20562#issuecomment-399100222>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWfoW6JF9F4iPRd7WO0m2Qlpbw8PTVuks5t-5z-gaJpZM4ULTfZ>
> .
>
-- 
- Michael Please pardon any errors, this message was sent from my iPhone.

dgrammatiko commented:
@infograf768 wasn't the answer from @mbabker here: https://github.com/joomla-projects/gsoc17_helpscreens_on_jdocs/pull/7#issuecomment-396570253 explanatory enough? Why are you keep crying over the freaking tooltips?

infograf768 commented:
> Why are you keep crying over the freaking tooltips?

Cool, my friend... Maybe just because nothing is done about it for the moment (that I have seen) and you proposed Note as a solution...

infograf768 commented:
@Hackwar 
I may have found an issue here and it is weird.
I have set the stemmer to use French on a monolanguage site. In my site is a phrase containing the word `française`. I can't get the suggestion. I checked that it is among the terms in db and it is.
If I set the stemmer to use English, I do get the suggestion OK, although the word is not highlighted in the result.

Stemmer set to english
<img width="505" alt="screen shot 2018-06-22 at 10 31 17" src="https://user-images.githubusercontent.com/869724/41766779-3325211e-7608-11e8-9500-0d7f176aa8b8.png">
result
<img width="887" alt="screen shot 2018-06-22 at 10 37 39" src="https://user-images.githubusercontent.com/869724/41766819-529b6a12-7608-11e8-9783-26519b02e6e7.png">

Did not test this yet on a multingual site.

carlitorweb commented:
@infograf768  is not from this PR. As I comment to @Hackwar  on [#20384 (comment)](https://github.com/joomla/joomla-cms/pull/20384#issuecomment-398077895) i think is some wrong JS.

Hackwar commented:
There is an issue with the stemmer setting. I'm going to fix that this evening. That it doesn't highlight the words is an issue with JS indeed. There are several issues with the JS and right now I would be happy if someone else could have a look and find solutions to those. My JS is not so strong...

carlitorweb commented:
Will be better open an issue, with all the JS problem you have found. In this way, others can have a look

infograf768 commented:
@carlitorweb 
the highlighting as well as the loss of focus on the field are indeed js related, but not the fact that a suggestion is found with English stemmer and not French stemmer. :)

carlitorweb commented:
@infograf768 that is true, sorry.

Hackwar committed:
Fixing option name

Hackwar committed:
Merge branch '4.0-dev' into j4finder_stemmonolingual

infograf768 commented:
@Hackwar 
No change versus https://github.com/joomla/joomla-cms/pull/20562#issuecomment-399367729

Hackwar committed:
Using the right language code for terms

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4finder_stemmonolingual

Hackwar committed:
Improving code for stem() and tokenize()

Hackwar commented:
Ok, so I found a few issues in the code. It works right for me now. But please make sure that you re-index before testing.

Hackwar commented:
What I mean is, that you have to recreate the index after you applied this patch.

infograf768 commented:
OK, this time it works better. I obtain suggestions. 👍 

Non ascii terms (forêt, française) are not highlighted in the results as we already know. Therefore I looked a bit further.

I dumped `$this->query->highlight` and this is what I get.
<img width="920" alt="screen shot 2018-06-27 at 08 21 56" src="https://user-images.githubusercontent.com/869724/41956236-324ba202-79e3-11e8-8fd8-c2e1c7b28492.png">

In this specific case, I solved the issue by modifying the highlighter method in `/libraries/cms/html/behavior.php`

I commented line 489 to 492
```
		//foreach ($terms as $i => $term)
		//{
		//	$terms[$i] = JFilterOutput::stringJSSafe($term);
		//}
```
<img width="902" alt="screen shot 2018-06-27 at 08 43 00" src="https://user-images.githubusercontent.com/869724/41957168-5798372a-79e6-11e8-99bf-93545c0c4572.png">

Therefore it looks like using `JFilterOutput::stringJSSafe` on non ascii creates this issue.
This is only used in the highlighter method in core.
That method loads the legacy/highligter.js. Could it be that that js does not interprete UTF8 correctly?

This issue is ALSO present in 3.x for com_finder. I guess it was never solved because the component is not much used.
The problem is NOT present when using com_search on 3.x.

@dgrammatiko

Hackwar commented:
Ok, so @infograf768 you have a successfull test, right? The JS issues are not related to this PR and they are a separate issue. (Which we indeed do have to fix for 4.0)

infograf768 commented:
I have tested this item :white_check_mark: successfully on fb32ca0ef5efc8040fa8c4c4f756e62f6f435390<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20562">issues.joomla.org/tracker/joomla-cms/20562</a>.</sub>

carlitorweb commented:
I have tested this item :white_check_mark: successfully on fb32ca0ef5efc8040fa8c4c4f756e62f6f435390<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20562">issues.joomla.org/tracker/joomla-cms/20562</a>.</sub>

franz-wohlkoenig commented:
Ready to Commit after two successful tests.

infograf768 commented:
@Hackwar 
Any way to show the `name="language_default"` field only when multilang is disabled?

Hackwar commented:
That field is especially necessary for non-multilang sites. So... No.Am 29.06.2018 16:14 schrieb infograf768 <notifications@github.com>:@Hackwar
Any way to show the name="language_default" field only when multilang is disabled?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

infograf768 commented:
> That field is especially necessary for non-multilang sites.

Sure that is why I asked if we could hide it when multilang is enabled. Could not find a way.

Hackwar commented:
You also need it for multilang for those texts that are marked *.Am 29.06.2018 16:34 schrieb infograf768 <notifications@github.com>:
That field is especially necessary for non-multilang sites.

Sure that is why I asked if we could hide it when multilang is enabled. Could not find a way.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

infograf768 commented:
> You also need it for multilang for those texts that are marked *

Correct. Thanks.

wilsonge commented:
Looks good to me! Thanks!
